### ISSUE TITLE
Internal dependencies organization

### ISSUE THREAD

**Issue by lemunozm** [centrifuge/centrifuge-chain#1380]
# Description
We don't have well-defined rules about what things should be where. Here I propose some changes and rules that I think can improve the repo organization and get rid of dependency problems in the future.

Feel free to discuss!

## Current state
Our repo has the following dependencies, and we have 3 main layers. We can have *n* `chains` using the same `runtime`, and we can have a `runtime` using *n* `pallets`.

![image](https://github.com/centrifuge/centrifuge-chain/assets/15687891/a350a5a6-005b-4afe-9234-dfdc7953963c)

Ideally, we would want to remove those red upward dependencies.
Down/right directions in this diagram mean "more generic", so those upper lines involve a lot of knowledge from a part that should not be known about. We only want to remain dependencies to "more generic" and avoid that generic things know about specific things.

## Desired
By removing the red lines, redefining the purpose of each crate, and adding a new one, we can get a down/right diagram:

These arrows would be transitive. It means i.e. that a *runtime* could make use of `cfg-utils`.

![image](https://github.com/centrifuge/centrifuge-chain/assets/15687891/be81d8b2-c62c-46bc-a3b2-37a0688c2155)

### Purpose of each crate:
From more generic (widely purpose) to less generic (centrifuge-related purpose):

- `cfg-utils`: Common utilities without any specific behavior associated. 
    - i.e. `Adjustement` type or `SECONDS_PER_MINUTE`, `DAYS` constant.
- `cfg-traits`: Common interfaces to all runtimes. It implies traits and also could imply types tied to those traits. But it should not use `cfg-types`, because they are specific types for configuring in the runtimes.
    - i.e. `Permission`, `InterestAccrual`, or `PoolInvest`.
- `cfg-types`: Should know about centrifuge-related types. These are types used in runtimes to configure pallets, but should not be used direclty in the pallets. They make sense for our use cases but not for others who can use our pallets. These types, therefore, should not know about pallets or runtimes. And should not be types used by `cfg-traits`.
    - i.e. complex types: `CurrencyId`, `FeeKey`, `Location` 
    - i.e. simple types: `PoolId`, `AccountId`, `Balance`. 
    - i.e. constants: `MAXIMUM_BLOCK_WEIGHT`, `CFG`(?), `SLOT_DURATION`, `TREASURY_FEE_RATIO`
    - If there are different types in each runtime, as `AIR` and `CFG`, I think each should be into some runtime-dependent module as `runtime/altair/src/types.rs`.
- `runtime-common`: Common utilities to all runtimes. It knows about how runtimes work and has hard dependencies on the pallet used. It's simply a common place where to store code common to all runtimes.
- `cfg-primitives`: Should know about constants and values needed for configuring chains, at the node level. Maybe the name could be changed to something more appropriate with this schema, or even be removed completely if there they are specified directly in `chain_spec.rs`.

**Comment by wischli** [comment]
Thanks for bringing this up and supporting your argument with nifty diagrams!

Say I have a custom type in `cfg-types` for which I want to implement a trait from `cfg-types`, e.g. some trait over our `CurrencyId` enum. Do I understand correctly, that this could only live in `runtime-common` if it is relevant on the runtime level?

**Comment by lemunozm** [comment]
Good question!

Probably, `cfg-types`, which lives in the centrifuge domain, can also access `cfg-traits` (and transitively to `cfg-primitives`), which is at a lower level.

Something like the blue line: 
![image](https://github.com/centrifuge/centrifuge-chain/assets/15687891/01e8f93e-7578-41f2-b4f5-fa91bd15c932)

**Comment by wischli** [comment]
So by applying this proposal, we would have to provide all `cfg-types` used on the pallet level via the `pallet::Config` trait, right?

**Comment by lemunozm** [comment]
Yes (that is what we are doing all the time). Or if the type is very generic, then it can be place in `cfg-primitives`, and let all pallets to use them directly.

**Comment by lemunozm** [comment]
@wischli @mustermeiszer I was thinking about if we would really need to differentiate between `runtime-common` and `cfg-types` in this schema, maybe both can be the same crate, WDYT? 🤔

**Comment by wischli** [comment]
> @wischli @mustermeiszer I was thinking about if we would really need to differentiate between `runtime-common` and `cfg-types` in this schema, maybe both can be the same crate, WDYT? 🤔

IMO, `cfg-types` could indeed be moved to a module inside `runtime-common` but I am sure @mustermeiszer has a stronger opinion.

**Comment by lemunozm** [comment]
I like what you say in your last comment @wischli, I think after this, the structure would be as follows:

![image](https://github.com/centrifuge/centrifuge-chain/assets/15687891/d791045a-ca86-4aa1-8e9e-a1387c4d66b3)

**Comment by lemunozm** [comment]
I think we need to keep this organization more than ever because right now:
`cfg-primitives` depends on `cumulus`, which depends on `xcm-builder` which seems to be broken if only `runtime-benchmarks` is enabled.

This makes that a lot of pallets can not be compiled using:
```sh
cargo test -p <pallet-name> --features runtime-benchmarks
```

and probably this is related to this issue too: https://github.com/centrifuge/centrifuge-chain/pull/1406

**Comment by mustermeiszer** [comment]
`cfg-primitives` are mostly primitives that are really specific to us. Mostly, all teams use the same setup -- e.g. `type Balance = u128` but no trait or pallet should depend on that IMO. But I agree with the above otherwise.

**Comment by lemunozm** [comment]
Updated diagram with the above:

![image](https://github.com/centrifuge/centrifuge-chain/assets/15687891/11147e0a-bf03-477e-83a1-2cc11a6016e6)

**Comment by mustermeiszer** [comment]
So `cfg-traits` MUST NOT contain any types from `cfg-types`?

**Comment by lemunozm** [comment]
Yes if those types are intended to be used to configure the runtime. 

Right now, `cfg_types` contains a mix of general purpose types and centrifuge-specific runtime types. The first one could be used by `cfg_traits` (as `Adjustement` type) but the second set of types should not. The first set of types as `Adjustement` could be placed in `cfg-utils`.

**Comment by lemunozm** [comment]
So, as an overview, we should divide `cfg-types` in:
- general purpose types, in `cfg_utils`.
- centrifuge-related types, in `runtime-common/types`
